ENH: to_sql() add parameter "method" to control insertions method (#8…

schettino72 opened the pull request:
…953)

- [x] closes #8953 
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

schettino72 committed:
ENH: to_sql() add parameter "method" to control insertions method (#8953)

pep8speaks commented:
Hello @schettino72! Thanks for submitting the PR.

 - In the file [`pandas/core/generic.py`](https://github.com/pandas-dev/pandas/blob/a77cdfd63df72c8b4d70783b12c0d9e4b748b9a7/pandas/core/generic.py), following are the PEP8 issues :

> [Line 2139:80](https://github.com/pandas-dev/pandas/blob/a77cdfd63df72c8b4d70783b12c0d9e4b748b9a7/pandas/core/generic.py#L2139): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (194 > 79 characters)

schettino72 commented:
- Added docs
- removed COPY implementation from code and added on docs

There is only one problem with flake8 complaining about a long.
The problem is that it is a url link, not sure how to break in into multiple lines...

schettino72 commented:
Note: I am targeting this to 0.24, not 0.23.x. 
IMO an addition of parameter should not be made a on bug release.
Otherwise code written for 0.32.2 using the new parameter would not run on 0.23.1...

schettino72 committed:
ENH: to_sql() add parameter "method". Fix docstrings (#8953)

gfyoung commented on pandas/core/generic.py:
```diff
@@ -2120,11 +2122,59 @@ def to_sql(self, name, con, schema=None, if_exists='fail', index=True,
+
+                  sql = 'COPY {} ({}) FROM STDIN WITH CSV'.format(
+                      table_name, columns)
+                  cur.copy_expert(sql=sql, file=s_buf)
```
Nice explanation here!  Do you mind moving some of it to the parameter description?  That's probably where viewers will look first.

jorisvandenbossche replied:
Yep, indeed (similar to my comment above), I would at least put a basic explanation of each option in the parameter description. 
And then you can refer to below for more detailed performance considerations / example ..

jorisvandenbossche replied:
And maybe we can put this example in the user guide (io.rst) ?

gfyoung commented:
> Note: I am targeting this to 0.24, not 0.23.x.

@schettino72 : That's exactly what we would do as well.  Good call!

gfyoung commented:
@schettino72 : Looks like you have a lint error according to Travis <a href="https://travis-ci.org/pandas-dev/pandas/jobs/390088220">here</a>/.

jorisvandenbossche commented on pandas/core/generic.py:
```diff
@@ -2052,6 +2052,8 @@ def to_sql(self, name, con, schema=None, if_exists='fail', index=True,
             names and the values should be the SQLAlchemy types or strings for
             the sqlite3 legacy mode.
+        method : {'default', 'multi', callable}, default 'default'
+            Controls the SQL insertion clause used.
```
I think we should give this a bit longer explanation

jorisvandenbossche replied:
And can you also add a `.. versionadded:: 0.24.0` here?

jorisvandenbossche commented on pandas/core/generic.py:
```diff
@@ -2120,11 +2122,59 @@ def to_sql(self, name, con, schema=None, if_exists='fail', index=True,
+            This usually provides a big performance for Analytic databases
+            like *Presto* and *Redshit*, but has worse performance for
+            traditional SQL backend if the table contains many columns.
+            For more information check SQLAlchemy `documention <http://docs.sqlalchemy.org/en/latest/core/dml.html?highlight=multivalues#sqlalchemy.sql.expression.Insert.values.params.*args>`__.
```
If you start the url on the next line, I think flake8 will be OK with the too long line, like:

```
    For more information check SQLAlchemy `documention 
    <http://docs.sqlalchemy.org/en/latest/core/dml.html?highlight=multivalues#sqlalchemy.sql.expression.Insert.values.params.*args>`__.
```

jorisvandenbossche commented on pandas/core/generic.py:
```diff
@@ -2120,11 +2122,59 @@ def to_sql(self, name, con, schema=None, if_exists='fail', index=True,
+
+         - `'default'`: Uses standard SQL `INSERT` clause
+         - `'multi'`: Pass multiple values in a single `INSERT` clause.
+            It uses a **special** SQL syntax not supported by all backends.
```
The 'It' should be indented equally to the `` `'multi'` `` of the line above (now it is one space more I think)

jorisvandenbossche commented on pandas/core/generic.py:
```diff
@@ -2120,11 +2122,59 @@ def to_sql(self, name, con, schema=None, if_exists='fail', index=True,
+            traditional SQL backend if the table contains many columns.
+            For more information check SQLAlchemy `documention <http://docs.sqlalchemy.org/en/latest/core/dml.html?highlight=multivalues#sqlalchemy.sql.expression.Insert.values.params.*args>`__.
+         - callable: with signature `(pd_table, conn, keys, data_iter)`.
+            This can be used to implement more performant insertion based on
```
same here about the indentation

jorisvandenbossche commented on pandas/core/generic.py:
```diff
@@ -2120,11 +2122,59 @@ def to_sql(self, name, con, schema=None, if_exists='fail', index=True,
+        The parameter ``method`` controls the SQL insertion clause used.
+        Possible values are:
+
+         - `'default'`: Uses standard SQL `INSERT` clause
```
Do we need to say this is a single insert statement for each row?

jorisvandenbossche commented on pandas/io/sql.py:
```diff
@@ -611,11 +635,18 @@ def insert_data(self):
+    def insert(self, chunksize=None, method=None):
+
+        # set insert method
+        if method in (None, 'default'):
```
is the 'None' needed here?

jorisvandenbossche reviewed (commented):
Thanks for the PR. Added some comments, didn't look at the tests yet

schettino72 committed:
ENH: to_sql() add parameter "method". Improve docs based on reviews (#8953)

schettino72 commented:
Thanks for quick review and tips. Ready for next round :grin:

gfyoung commented:
@schettino72 : Seems like you have test failures...

schettino72 committed:
ENH: to_sql() add parameter "method". Fix unit-test (#8953)

jreback commented:
lgtm. @jorisvandenbossche merge when satisfied.

jreback commented:
lgtm. ping @jorisvandenbossche

dahlbaek commented on doc/source/io.rst:
```diff
@@ -4752,6 +4752,57 @@ default ``Text`` type for string columns:
+              table_name = table.name
+
+          sql = 'COPY {} ({}) FROM STDIN WITH CSV'.format(
+              table_name, columns)
```
I realise this is just an example, but isn't this piece vulnerable to SQL injection? the[`psycopg2.sql`](http://initd.org/psycopg/docs/sql.html) module has an [`Identifier`](http://initd.org/psycopg/docs/sql.html#psycopg2.sql.Identifier) object designed to handle SQL identifiers appropriately. See also [Example 41-1. Quoting Values In Dynamic Queries](https://www.postgresql.org/docs/9.6/static/plpgsql-statements.html#PLPGSQL-QUOTE-LITERAL-EXAMPLE).

jorisvandenbossche replied:
@dahlbaek no expert on this, but if you could provide the code that you would replace it with to make it more robust, that would certainly be welcome!

danfrankj commented:
@schettino72 would love to see this this ship - think we'll get it into the next release?

TomAugspurger commented:
What's the status on this?

@schettino72 there's a merge conflict with master, if you could merge master and push.

wxianxin commented:
@schettino72 The magnificent you are the only hope for this unbelievable feature to be delivered at this wonderful release.

schettino72 commented:
Sorry, I dont have the time now to work on this.

It seems the conflict is only on README file.
IMO should be resolved at the time the pull request is actually merged (by someone with write access), not something i can do.
Otherwise I might be playing catch up forever...

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into schettino72-tosql-insert-param-8953

jorisvandenbossche commented:
The bottleneck is actually me (or another core dev) doing a new review, or if we think it is not needed it should be ready to merge if the merge conflicts is fixed.

But taking another look now.

TomAugspurger commented:
Merged master.

On Tue, Oct 23, 2018 at 2:13 PM eduardo naufel schettino <
notifications@github.com> wrote:

> Sorry, I dont have the time now to work on this.
>
> It seems the conflict is only on README file.
> IMO should be resolved at the time the pull request is actually merged (by
> someone with write access), not something i can do.
> Otherwise I might be playing catch up forever...
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/21401#issuecomment-432381507>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABQHIi9VCOM3XtPLrmWDWOsN0krDWKvhks5un2pvgaJpZM4UhSR4>
> .
>

jorisvandenbossche committed:
doc clean-up

jorisvandenbossche commented on doc/source/io.rst:
```diff
@@ -4806,6 +4806,55 @@ default ``Text`` type for string columns:
+The parameter ``method`` controls the SQL insertion clause used.
+Possible values are:
+
+- ``'default'``: Uses standard SQL ``INSERT`` clause (one per row).
```
The "one per row", is that correct? 

Because as far as I understand, the default behaviour will be to do a single insert statement with multiple rows (a "executemany"), where the number of rows is determined by the chunksize (and which by default is all rows)

schettino72 replied:
> The "one per row", is that correct?

Maybe that is why everybody is complaining that it is so slow :grin:

jorisvandenbossche commented on doc/source/io.rst:
```diff
@@ -4806,6 +4806,55 @@ default ``Text`` type for string columns:
+          s_buf = StringIO()
+          writer = csv.writer(s_buf)
+          writer.writerows(data_iter)
+          s_buf.seek(0)
```
Is there a reason you are not using `to_csv` here? It should be possible to write to a buffer with that as well.

Further, I think it would be good to actually run the code of this example, to make sure it works. We could either add it as a test case, or either actually run it in the documentation build (by applying it on a small example dataframe)

schettino72 replied:
My understanding is that `to_csv()` is a higher level function do a bunch of data conversions. 
At this point when this function is called `data_iter` has already passed trough pandas data conversions...

TomAugspurger replied:
Are we fixing this before merging? I can push a commit if we want.

TomAugspurger replied:
Started out as that, but this is kind of a can of worms to actually run this example. We would need to actually create a Postgres database. I'll see if something similar exists for sqlite.

TomAugspurger replied:
Actually nevermind, this is tested below and we don't actually have a DataFrame here we have a `chunk_iter`. It's not clear to me, but that seems to be a list of arrays, one per column, chunk pair?

TomAugspurger replied:
So +1 to merging as is.

jorisvandenbossche committed:
additional doc clean-up

jreback commented:
@schettino72 can you merge master

tgriek commented:
Waiting for this! Thanks guys.

Commit:
Merge remote-tracking branch 'upstream/master' into tosql-insert-param-8953

Commit:
use dict(zip()) directly

Commit:
clean up merge

jreback commented on pandas/io/sql.py:
```diff
@@ -395,7 +396,7 @@ def read_sql(sql, con, index_col=None, coerce_float=True, params=None,
 
 def to_sql(frame, name, con, schema=None, if_exists='fail', index=True,
-           index_label=None, chunksize=None, dtype=None):
+           index_label=None, chunksize=None, dtype=None, method='default'):
```
``method=None`` would be better here

Commit:
default --> None

Commit:
Remove stray default

Commit:
Remove method kwarg

Commit:
change default to None

Commit:
test copy insert snippit

Commit:
print debug

Commit:
index=False

jreback commented on doc/source/whatsnew/v0.24.0.rst:
```diff
@@ -373,6 +373,7 @@ Other Enhancements
 - :meth:`DataFrame.between_time` and :meth:`DataFrame.at_time` have gained the an ``axis`` parameter (:issue:`8839`)
 - The ``scatter_matrix``, ``andrews_curves``, ``parallel_coordinates``, ``lag_plot``, ``autocorrelation_plot``, ``bootstrap_plot``, and ``radviz`` plots from the ``pandas.plotting`` module are now accessible from calling :meth:`DataFrame.plot` (:issue:`11978`)
 - :class:`IntervalIndex` has gained the :attr:`~IntervalIndex.is_overlapping` attribute to indicate if the ``IntervalIndex`` contains any overlapping intervals (:issue:`23309`)
+- :func:`pandas.DataFrame.to_sql` has gained the ``method`` argument to control SQL insertion clause (:issue:`8953`)
```
can you add a pointer to the new doc section?

mroeschke replied:
Done.

jreback commented on pandas/io/sql.py:
```diff
@@ -1115,7 +1156,17 @@ def to_sql(self, frame, name, if_exists='fail', index=True,
+            Controls the SQL insertion clause used:
+
+            * None : Uses standard SQL ``INSERT`` clause (one per row).
+            * 'multi': Pass multiple values in a single ``INSERT`` clause.
```
seems to be lots of repetitive doc-strings here, can think about this for later

mroeschke replied:
Sure, we can probably template this a later PR.

jreback reviewed (changes requested):
lgtm. minor comment on the whatsnew.

Commit:
Add reference to documentation

jreback commented:
@TomAugspurger ?

mroeschke commented:
Thanks @jreback, @TomAugspurger.

And thank you @schettino72!

danfrankj commented:
Thank you all! Looking forward to using this :)

maxgrenderjones commented:
Hooray! Thanks :)
